File: uit.hotel/Models/Price/PriceItem.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using Realms;
using uit.hotel.Queries.Helper;

namespace uit.hotel.Models
{
    public enum PriceItemKindEnum
    {
        [Description("Giá theo giờ")] Hour,
        [Description("Giá theo đêm")] Night,
        [Description("Giá theo ngày")] Day,
        [Description("Giá theo tuần")] Week,
        [Description("Giá theo tháng")] Month
    }

    public class PriceItem : RealmObject
    {
        public Booking Booking { get; set; }

        [Ignored]
        public TimeSpan TimeSpan { get => new TimeSpan(TimeSpanRaw); set { TimeSpanRaw = value.Ticks; } }
        private long TimeSpanRaw { get; set; }

        [Ignored]
        public PriceItemKindEnum Kind { get => (PriceItemKindEnum)KindRaw; set { KindRaw = (int)value; } }
        private int KindRaw { get; set; }

        // Calculated Value
        [Ignored]
        public long Value
        {
            get
            {
                switch (Kind)
                {
                    case PriceItemKindEnum.Hour:
                        return TimeSpan.MultiplyByHourPrice(Booking.Price.HourPrice);
                    case PriceItemKindEnum.Night:
                        return Booking.Price.NightPrice;
                    case PriceItemKindEnum.Day:
                        return Booking.Price.DayPrice * (TimeSpan.Days + 1);
                    case PriceItemKindEnum.Week:
                        return Booking.Price.WeekPrice * ((TimeSpan.Days + 1) / 7);
                    case PriceItemKindEnum.Month:
                        return Booking.Price.MonthPrice * ((TimeSpan.Days + 1) / 30);
                }
                return 0;
            }
        }

        public PriceItem Set(PriceItemKindEnum kind, TimeSpan timeSpan)
        {
            Kind = kind;
            TimeSpan = timeSpan;
            return this;
        }
    }
}

File: uit.hotel.client/modules/model/src/priceItem.ts
import { duration } from 'moment';
import { GetBooking, PriceItemKindEnum } from '~/graphql/types';
import { toMoney } from '~/utils';

export class PriceItemObject {
    price: GetBooking.Price;
    priceItem: GetBooking.PriceItems;

    constructor(price: GetBooking.Price, priceItem: GetBooking.PriceItems) {
        this.price = price;
        this.priceItem = priceItem;
    }

    get amount() {
        const { timeSpan, kind } = this.priceItem;
        const time = duration(timeSpan, 'second');
        return parseFloat(
            {
                [PriceItemKindEnum.Hour]: () => time.asHours(),
                [PriceItemKindEnum.Night]: () => 1,
                [PriceItemKindEnum.Day]: () => time.add(2, 'hour').asDays(),
                [PriceItemKindEnum.Week]: () => time.add(2, 'hour').asWeeks(),
                [PriceItemKindEnum.Month]: () => time.add(2, 'hour').asMonths(),
            }
                [kind]()
                .toFixed(2),
        );
    }

    get unit() {
        const { kind } = this.priceItem;
        return {
            [PriceItemKindEnum.Hour]: 'giờ',
            [PriceItemKindEnum.Night]: 'đêm',
            [PriceItemKindEnum.Day]: 'ngày',
            [PriceItemKindEnum.Week]: 'tuần',
            [PriceItemKindEnum.Month]: 'tháng',
        }[kind];
    }

    get unitPrice() {
        const { price, priceItem } = this;
        return {
            [PriceItemKindEnum.Hour]: () => price.hourPrice,
            [PriceItemKindEnum.Night]: () => price.nightPrice,
            [PriceItemKindEnum.Day]: () => price.dayPrice,
            [PriceItemKindEnum.Week]: () => price.weekPrice,
            [PriceItemKindEnum.Month]: () => price.monthPrice,
        }[priceItem.kind]();
    }

    get total() {
        return this.priceItem.value;
    }

    get number() {
        return `${this.amount} ${this.unit}`;
    }

    get text() {
        const { unit, unitPrice, amount } = this;
        return `${amount} ${unit} ✕ ${toMoney(unitPrice)}`;
    }
}
